src/lls_model.jl

```diff
@@ -1,27 +1,165 @@
-export LLSModel
+export LLSMatrixModel, LLSOperatorModel, LLSTripletModel, LLSModel, LLSModels
 
-"""
-    nls = LLSModel(A, b; lvar, uvar, C, lcon, ucon)
+abstract type LLSModels <: AbstractNLSModel end
 
-Creates a Linear Least Squares model ``\\tfrac{1}{2}\\|Ax - b\\|^2`` with optional bounds
-`lvar ≦ x ≦ uvar` and optional linear constraints `lcon ≦ Cx ≦ ucon`.
-This problem is a nonlinear least-squares problem with residual given by ``F(x) = Ax - b``.
-"""
-mutable struct LLSModel <: AbstractNLSModel
+mutable struct LLSMatrixModel <: LLSModels
+  meta :: NLPModelMeta
+  nls_meta :: NLSMeta
+  counters :: NLSCounters
+
+  A
+  b :: Vector
+  C
+end
+
+mutable struct LLSOperatorModel <: LLSModels
+  meta :: NLPModelMeta
+  nls_meta :: NLSMeta
+  counters :: NLSCounters
+
+  A :: LinearOperator
+  b :: Vector
+  C :: LinearOperator
+end
+
+mutable struct LLSTripletModel <: LLSModels
   meta :: NLPModelMeta
   nls_meta :: NLSMeta
   counters :: NLSCounters
 
   Arows :: Vector{Int}
   Acols :: Vector{Int}
   Avals :: Vector
-  b :: AbstractVector
+  b :: Vector
   Crows :: Vector{Int}
   Ccols :: Vector{Int}
   Cvals :: Vector
 end
 
+"""
+    nls = LLSModel(A, b; lvar, uvar, C, lcon, ucon, variant)
+    nls = LLSModel(opA, b; lvar, uvar, opC, lcon, ucon)
+    nls = LLSModel(Arows, Acols, Avals, b; lvar, uvar, Crows, Ccols, Cvals, lcon, ucon, variant)
+
+Creates a Linear Least Squares model ``\\tfrac{1}{2}\\|Ax - b\\|^2`` with optional bounds
+`lvar ≦ x ≦ uvar` and optional linear constraints `lcon ≦ Cx ≦ ucon`.
+This problem is a nonlinear least-squares problem with residual given by ``F(x) = Ax - b``.
+
+There are three internal structures for `LLSModels`, one storing the matrices directly, the
+other storing `LinearOperator`s, and the last storing the triplet format of the matrices.
+They correspond, respectively, for each of the three function calls shown.
+Additionally, the `variant` can be used to transform the input into a differnt internal structure.
+The option can be `:matrix`, `:operator`, and `:triplet`. Notice that the function signature
+for LinearOperators input does not accept the variant option.
+"""
+function LLSModel end
+
+function LLSMatrixModel(A :: AbstractMatrix, b :: AbstractVector;
```

Comment:
Could all the constructors be called `LLSModel`?

Reply:
I have constructors called `LLSModel` with the use of `Val(variant)`, and I have these as the underlying constructor. Do you still want to remove these?